File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/UseSmileIDSampleTestIds.kt
package com.usesmileid.sampleapps.ui

/** The accessibility ids this module attaches. Stable forever: deprecate, never rename. */
object UseSmileIDSampleTestIds {
    const val NAV_PRODUCTS = "sample_nav_products"
    const val NAV_VERIFICATIONS = "sample_nav_verifications"
    const val NAV_SETTINGS = "sample_nav_settings"
    const val NAV_TOKEN = "sample_nav_token"

    const val PRODUCTS_SCREEN = "sample_products_screen"
    const val VERIFICATIONS_SCREEN = "sample_verifications_screen"
    const val VERIFICATIONS_EMPTY = "sample_verifications_empty"
    const val DETAILS_EMPTY = "sample_details_empty"
    const val COUNTRY_EMPTY = "sample_country_empty"
    const val ID_TYPE_EMPTY = "sample_idtype_empty"
    const val VERIFICATION_DETAILS_SCREEN = "sample_verification_details_screen"
    const val SETTINGS_SCREEN = "sample_settings_screen"
    const val USER_DETAILS_SCREEN = "sample_user_details_screen"
    const val KYC_FORM_SCREEN = "sample_kyc_form_screen"
    const val COUNTRY_SHEET = "sample_country_sheet"
    const val ID_TYPE_SHEET = "sample_idtype_sheet"
    const val PROFILE_SWITCH_SHEET = "sample_profile_switch_sheet"
    const val PROFILES_SCREEN = "sample_profiles_screen"
    const val PROFILE_CONFIG_SCREEN = "sample_profile_config_screen"
    const val NEW_PROFILE_SHEET = "sample_new_profile_sheet"
    const val SCAN_TOKEN_SCREEN = "sample_scan_token_screen"

    // The drawer's own root; its opener is a separate id, matching every other sheet and opener.
    const val SCENARIO_DRAWER = "sample_scenario_drawer"
    const val SCENARIO_DRAWER_BUTTON = "sample_scenario_drawer_button"

    const val PROFILE_SUMMARY = "sample_profile_summary"
    const val SETTING_SMILE_TO_CAPTURE = "sample_setting_smile_to_capture"
    const val SETTING_AGENT_MODE = "sample_setting_agent_mode"
    const val SETTING_DARK_MODE = "sample_setting_dark_mode"
    const val SETTING_CONSENT_STEP = "sample_setting_consent_step"
    const val SETTING_INSTRUCTIONS_STEP = "sample_setting_instructions_step"
    const val SETTING_PREVIEW_STEP = "sample_setting_preview_step"
    const val SIGN_OUT = "sample_sign_out"
    const val VERSION_LABEL = "sample_version_label"
    const val SELECT_TOGGLE = "sample_select_toggle"
    const val STATUS_BADGE = "sample_status_badge"
    const val DETAILS_REFRESH = "sample_details_refresh"
    const val DETAILS_DELETE = "sample_details_delete"
    const val USER_DETAILS_HINT = "sample_user_details_hint"
    const val REMEMBER_DETAILS_SWITCH = "sample_remember_details_switch"
    const val USER_DETAILS_CONTINUE = "sample_user_details_continue"
    const val COUNTRY_TRIGGER = "sample_country_trigger"
    const val ID_TYPE_TRIGGER = "sample_idtype_trigger"
    const val ID_NUMBER_INPUT = "sample_idnumber_input"
    const val KYC_CONTINUE = "sample_kyc_continue"
    const val COUNTRY_SEARCH = "sample_country_search"
    const val ID_TYPE_SEARCH = "sample_idtype_search"
    const val CREATE_PROFILE = "sample_create_profile"
    const val NEW_PROFILE_NAME = "sample_new_profile_name"
    const val NEW_PROFILE_FIRST_NAME = "sample_new_profile_first_name"
    const val NEW_PROFILE_LAST_NAME = "sample_new_profile_last_name"
    const val NEW_PROFILE_EMAIL = "sample_new_profile_email"
    const val NEW_PROFILE_PHONE = "sample_new_profile_phone"
    const val NEW_PROFILE_SAVE = "sample_new_profile_save"
    const val PROFILE_CONFIG_SAVE = "sample_profile_config_save"

    // Attached by the component itself: the spec assigns these to it, not to a screen.
    const val TOAST = "sample_toast"
    const val TOAST_UNDO = "sample_toast_undo"
    const val SELECTION_BAR = "sample_selection_bar"
    const val SELECTION_COUNT = "sample_selection_count"
    const val SELECTION_REMOVE = "sample_selection_remove"
    const val JOB_ROW_STATUS = "sample_job_row_status"
    const val SESSION_CARD = "sample_session_card"
    const val SESSION_COUNTDOWN = "sample_session_countdown"
    const val SESSION_ENDED_BANNER = "sample_session_ended_banner"
    const val ENV_CHIP = "sample_env_chip"
    const val PROFILE_AVATAR_BUTTON = "sample_profile_avatar_button"
    const val TOKEN_FLOAT = "sample_token_float"
    const val TOKEN_MANUAL_ENTRY = "sample_token_manual_entry"
    const val TOKEN_PASTE = "sample_token_paste"
    const val TOKEN_SIMULATE = "sample_token_simulate"

    const val RESULT_CARD = "sample_result_card"
    const val RESULT_ACTIVE_SCENARIO = "sample_result_active_scenario"
    const val RESULT_ACTIVE_THEME = "sample_result_active_theme"
    const val RESULT_ROUTE = "sample_result_route"
    const val RESULT_ENVIRONMENT = "sample_result_environment"
    const val RESULT_JOB_ID = "sample_result_job_id"
    const val RESULT_USER_ID = "sample_result_user_id"
    const val RESULT_JOB_STATUS = "sample_result_job_status"
    const val RESULT_RESULT_COUNT = "sample_result_result_count"
    const val RESULT_REFRESH_COUNT = "sample_result_refresh_count"
    const val RESULT_LAST_ERROR = "sample_result_last_error"
    const val RESULT_SDK_VERSION = "sample_result_sdk_version"

    val all = listOf(
        NAV_PRODUCTS,
        NAV_VERIFICATIONS,
        NAV_SETTINGS,
        NAV_TOKEN,
        PRODUCTS_SCREEN,
        VERIFICATIONS_SCREEN,
        VERIFICATIONS_EMPTY,
        DETAILS_EMPTY,
        COUNTRY_EMPTY,
        ID_TYPE_EMPTY,
        VERIFICATION_DETAILS_SCREEN,
        SETTINGS_SCREEN,
        USER_DETAILS_SCREEN,
        KYC_FORM_SCREEN,
        COUNTRY_SHEET,
        ID_TYPE_SHEET,
        PROFILE_SWITCH_SHEET,
        PROFILES_SCREEN,
        PROFILE_CONFIG_SCREEN,
        NEW_PROFILE_SHEET,
        SCAN_TOKEN_SCREEN,
        SCENARIO_DRAWER,
        TOAST,
        TOAST_UNDO,
        SELECTION_BAR,
        SELECTION_COUNT,
        SELECTION_REMOVE,
        JOB_ROW_STATUS,
        SESSION_CARD,
        SESSION_COUNTDOWN,
        SESSION_ENDED_BANNER,
        ENV_CHIP,
        PROFILE_AVATAR_BUTTON,
        TOKEN_FLOAT,
        TOKEN_MANUAL_ENTRY,
        TOKEN_PASTE,
        TOKEN_SIMULATE,
        SCENARIO_DRAWER_BUTTON,
        PROFILE_SUMMARY,
        SETTING_SMILE_TO_CAPTURE,
        SETTING_AGENT_MODE,
        SETTING_DARK_MODE,
        SETTING_CONSENT_STEP,
        SETTING_INSTRUCTIONS_STEP,
        SETTING_PREVIEW_STEP,
        SIGN_OUT,
        VERSION_LABEL,
        SELECT_TOGGLE,
        STATUS_BADGE,
        DETAILS_REFRESH,
        DETAILS_DELETE,
        USER_DETAILS_HINT,
        REMEMBER_DETAILS_SWITCH,
        USER_DETAILS_CONTINUE,
        COUNTRY_TRIGGER,
        ID_TYPE_TRIGGER,
        ID_NUMBER_INPUT,
        KYC_CONTINUE,
        COUNTRY_SEARCH,
        ID_TYPE_SEARCH,
        CREATE_PROFILE,
        NEW_PROFILE_NAME,
        NEW_PROFILE_FIRST_NAME,
        NEW_PROFILE_LAST_NAME,
        NEW_PROFILE_EMAIL,
        NEW_PROFILE_PHONE,
        NEW_PROFILE_SAVE,
        PROFILE_CONFIG_SAVE,
        RESULT_CARD,
        RESULT_ACTIVE_SCENARIO,
        RESULT_ACTIVE_THEME,
        RESULT_ROUTE,
        RESULT_ENVIRONMENT,
        RESULT_JOB_ID,
        RESULT_USER_ID,
        RESULT_JOB_STATUS,
        RESULT_RESULT_COUNT,
        RESULT_REFRESH_COUNT,
        RESULT_LAST_ERROR,
        RESULT_SDK_VERSION,
    )

    /** Suffixed ids derive from other spec files, so they are built rather than listed. */
    fun productCard(productId: String) = "sample_product_card_$productId"

    fun settingNav(rowId: String) = "sample_setting_nav_$rowId"

    fun scenarioItem(scenarioId: String) = "sample_scenario_item_$scenarioId"

    fun themeItem(themeId: String) = "sample_theme_item_$themeId"

    fun filterChip(status: String) = "sample_filter_chip_$status"

    fun filterCount(status: String) = "sample_filter_count_$status"

    fun jobRow(index: Int) = "sample_job_row_$index"

    fun selectionCheckbox(index: Int) = "sample_selection_checkbox_$index"

    fun detailField(field: String) = "sample_detail_field_$field"

    fun detailCopy(field: String) = "sample_detail_copy_$field"

    fun userDetailsField(field: String) = "sample_user_details_field_$field"

    fun countryOption(isoCode: String) = "sample_country_option_$isoCode"

    fun idTypeOption(typeId: String) = "sample_idtype_option_$typeId"

    fun profileRow(profileId: String) = "sample_profile_row_$profileId"

    fun profileConfigField(field: String) = "sample_profile_config_field_$field"
}

File: flutter/sample_ui/lib/src/use_smileid_sample_test_ids.dart
/// A coordinate carries characters an id may not, folded the same way on every platform.
String _idSafe(String value) => value.replaceAll(RegExp('[^A-Za-z0-9]'), '_');

/// The accessibility ids this package attaches. Stable forever: deprecate, never rename.
abstract final class UseSmileIDSampleTestIds {
  /// Bottom nav destination: products.
  static const String navProducts = 'sample_nav_products';

  /// Bottom nav destination: verifications.
  static const String navVerifications = 'sample_nav_verifications';

  /// Bottom nav destination: settings.
  static const String navSettings = 'sample_nav_settings';

  /// The bottom nav's token affordance, which carries the countdown ring.
  static const String navToken = 'sample_nav_token';

  /// The active-session card on the products screen.
  static const String sessionCard = 'sample_session_card';

  /// The countdown inside the session card, asserted as a value.
  static const String sessionCountdown = 'sample_session_countdown';

  /// The neutral card that replaces the session card on expiry.
  static const String sessionEndedBanner = 'sample_session_ended_banner';

  /// The floating affordance that reaches the token session from inside a form.
  static const String tokenFloat = 'sample_token_float';

  /// The manual token entry row on the scan screen.
  static const String tokenManualEntry = 'sample_token_manual_entry';

  /// The paste action inside the manual entry row.
  static const String tokenPaste = 'sample_token_paste';

  /// Simulate a successful scan, which makes token flows testable with no QR source.
  static const String tokenSimulate = 'sample_token_simulate';

  /// The scan token screen root.
  static const String scanTokenScreen = 'sample_scan_token_screen';

  /// One environment chip, suffixed with the environment id.
  static String tokenEnvironment(String environmentId) =>
      'sample_token_environment_$environmentId';

  /// The products screen root.
  static const String productsScreen = 'sample_products_screen';

  /// The header avatar button, which opens the profile-switch sheet.
  static const String profileAvatarButton = 'sample_profile_avatar_button';

  /// One product card, suffixed with the product id.
  static String productCard(String productId) =>
      'sample_product_card_$productId';

  /// The verifications root.
  static const String verificationsScreen = 'sample_verifications_screen';

  /// The verifications list with nothing in it, which is what a first launch shows.
  static const String verificationsEmpty = 'sample_verifications_empty';

  /// The settings root.
  static const String settingsScreen = 'sample_settings_screen';

  /// The profile row at the top of settings, which opens the profiles LIST.
  static const String profileSummary = 'sample_profile_summary';

  /// The Enhanced SmartSelfie switch, which drives enableEnhancedLiveness.
  static const String settingEnhancedSmartSelfie =
      'sample_setting_enhanced_smart_selfie';

  /// The Agent mode switch, which drives allowAgentMode.
  static const String settingAgentMode = 'sample_setting_agent_mode';

  /// The Dark mode switch.
  static const String settingDarkMode = 'sample_setting_dark_mode';

  /// The Consent screen switch, which includes or omits consent().
  static const String settingConsentStep = 'sample_setting_consent_step';

  /// The Instruction screen switch, which includes or omits instructions().
  static const String settingInstructionsStep =
      'sample_setting_instructions_step';

  /// The Preview screen switch, which includes or omits preview().
  static const String settingPreviewStep = 'sample_setting_preview_step';

  /// One ABOUT or LEGAL navigation row; suffixed with the row id.
  static const String settingNav = 'sample_setting_nav';

  /// The sign-out row.
  static const String signOut = 'sample_sign_out';

  /// The sign-out confirmation's destructive action.
  static const String signOutConfirm = 'sample_sign_out_confirm';

  /// The app name and version footer.
  static const String versionLabel = 'sample_version_label';

  /// The toast container, shared by the removal confirmation and the profile-created one.
  static const String toast = 'sample_toast';

  /// The action inside a toast, whatever it is labelled.
  static const String toastUndo = 'sample_toast_undo';

  /// One filter chip, suffixed with the filter's id.
  static String filterChip(String filterId) => 'sample_filter_chip_$filterId';

  /// A filter chip's count, asserted separately so a flow reads a number rather than prose.
  static String filterCount(String filterId) => 'sample_filter_count_$filterId';

  /// One verification row, suffixed with its position in the whole visible list.
  static String jobRow(int index) => 'sample_job_row_$index';

  /// The status badge inside a verification row, which the row attaches itself.
  static const String jobRowStatus = 'sample_job_row_status';

  /// The header action that enters and leaves select mode; its label flips Select / Cancel.
  static const String selectToggle = 'sample_select_toggle';

  /// One row's selection checkbox, suffixed with the row's position, present only in select mode.
  static String selectionCheckbox(int index) =>
      'sample_selection_checkbox_$index';

  /// One verification's detail page.
  static const String verificationDetailsScreen =
      'sample_verification_details_screen';

  /// The status badge on the detail page, which is NOT the row's badge id.
  static const String statusBadge = 'sample_status_badge';

  /// One label/value row on the detail page, suffixed with the field name.
  static String detailField(String field) => 'sample_detail_field_$field';

  /// A detail row's copy control, suffixed with the same field name.
  static String detailCopy(String field) => 'sample_detail_copy_$field';

  /// The detail page's destructive action, which hides the verification.
  static const String detailsDelete = 'sample_details_delete';

  /// The pull-to-refresh container, which the design draws in the processing state.
  static const String detailsRefresh = 'sample_details_refresh';

  /// What the detail page shows for a job this build never stored.
  static const String detailsEmpty = 'sample_details_empty';

  /// The profiles list.
  static const String profilesScreen = 'sample_profiles_screen';

  /// One profile row, suffixed with that profile's id; the list and the switch sheet share it.
  static String profileRow(String profileId) => 'sample_profile_row_$profileId';

  /// The row that opens the new-profile sheet.
  static const String createProfile = 'sample_create_profile';

  /// One profile's own page.
  static const String profileConfigScreen = 'sample_profile_config_screen';

  /// One editable field on that page, suffixed with the field's camelCase id.
  static String profileConfigField(String field) =>
      'sample_profile_config_field_$field';

  /// The profile's webhook URL, in its own section below the details.
  static const String profileConfigCallbackUrl =
      'sample_profile_config_callback_url';

  /// The page's only write: it saves the details AND makes the profile active.
  static const String profileConfigSave = 'sample_profile_config_save';

  /// The profile's organisation row.
  static const String profileConfigName = 'sample_profile_config_name';

  /// Deletes the profile after a confirmation.
  static const String profileConfigDelete = 'sample_profile_config_delete';

  /// The delete confirmation's destructive action.
  static const String profileDeleteConfirm = 'sample_profile_delete_confirm';

  /// The result card container, and the compact line that stands in for it on products.
  static const String resultCard = 'sample_result_card';

  /// The scenario the run actually got.
  static const String resultActiveScenario = 'sample_result_active_scenario';

  /// The theme scenario in effect.
  static const String resultActiveTheme = 'sample_result_active_theme';

  /// The route hosting the flow.
  static const String resultRoute = 'sample_result_route';

  /// The environment the run submitted to; the only surface that says so.
  static const String resultEnvironment = 'sample_result_environment';

  /// The SDK's job id.
  static const String resultJobId = 'sample_result_job_id';

  /// The server's user id.
  static const String resultUserId = 'sample_result_user_id';

  /// The run status, cancelled and failed kept apart.
  static const String resultJobStatus = 'sample_result_job_status';

  /// How many times the host result callback fired.
  static const String resultResultCount = 'sample_result_result_count';

  /// How many times the token refresh callback fired.
  static const String resultRefreshCount = 'sample_result_refresh_count';

  /// The last error the host saw.
  static const String resultLastError = 'sample_result_last_error';

  /// The SDK version the run exercised.
  static const String resultSdkVersion = 'sample_result_sdk_version';

  /// The switch sheet, which Products owns.
  static const String profileSwitchSheet = 'sample_profile_switch_sheet';

  /// The switch sheet's last row, which creates a profile and makes it active.
  static const String profileSwitchNew = 'sample_profile_switch_new';

  /// The new-profile sheet, which the profiles list owns.
  static const String newProfileSheet = 'sample_new_profile_sheet';

  /// The new profile's organisation.
  static const String newProfileName = 'sample_new_profile_name';

  /// The new profile's given name.
  static const String newProfileFirstName = 'sample_new_profile_first_name';

  /// The new profile's family name.
  static const String newProfileLastName = 'sample_new_profile_last_name';

  /// The new profile's email, which never gates the confirm.
  static const String newProfileEmail = 'sample_new_profile_email';

  /// The new profile's phone, which never gates the confirm.
  static const String newProfilePhone = 'sample_new_profile_phone';

  /// Creates the profile; the confirmation offers to make it active.
  static const String newProfileSave = 'sample_new_profile_save';

  /// The consent details form, which every product shows before its flow.
  static const String userDetailsScreen = 'sample_user_details_screen';

  /// One of its four rows, suffixed with the field's camelCase id.
  static String userDetailsField(String field) =>
      'sample_user_details_field_$field';

  /// The line under the card that says what is still outstanding.
  static const String userDetailsHint = 'sample_user_details_hint';

  /// The remember switch, which appears only once the form is satisfied.
  static const String rememberDetailsSwitch = 'sample_remember_details_switch';

  /// Whose details the form shows; opens the switch sheet.
  static const String userDetailsProfile = 'sample_user_details_profile';

  /// The consent form's continue.
  static const String userDetailsContinue = 'sample_user_details_continue';

  /// The ID details form, shown only for the products that need one.
  static const String kycFormScreen = 'sample_kyc_form_screen';

  /// The country select trigger.
  static const String countryTrigger = 'sample_country_trigger';

  /// The ID type select trigger, disabled until a country is chosen.
  static const String idTypeTrigger = 'sample_idtype_trigger';

  /// The ID number field.
  static const String idNumberInput = 'sample_idnumber_input';

  /// The ID form's continue.
  static const String kycContinue = 'sample_kyc_continue';

  /// The country picker sheet.
  static const String countrySheet = 'sample_country_sheet';

  /// Its search field, which filters on the country NAME and never the code.
  static const String countrySearch = 'sample_country_search';

  /// One country row, suffixed with its ISO code.
  static String countryOption(String code) => 'sample_country_option_$code';

  /// What the country picker shows when the search matches nothing.
  static const String countryEmpty = 'sample_country_empty';

  /// The ID type picker sheet.
  static const String idTypeSheet = 'sample_idtype_sheet';

  /// Its search field.
  static const String idTypeSearch = 'sample_idtype_search';

  /// One ID type row, suffixed with its id.
  static String idTypeOption(String typeId) => 'sample_idtype_option_$typeId';

  /// What the ID type picker shows with no country, and with no match.
  static const String idTypeEmpty = 'sample_idtype_empty';

  /// The scenario drawer, a debug affordance the design does not cover.
  static const String scenarioDrawer = 'sample_scenario_drawer';

  /// The settings row that opens it.
  static const String scenarioDrawerButton = 'sample_scenario_drawer_button';

  /// One flow-scenario row, suffixed with the scenario id.
  static String scenarioItem(String scenarioId) =>
      'sample_scenario_item_$scenarioId';

  /// One theme-scenario row, suffixed with the theme id.
  static String themeItem(String themeId) => 'sample_theme_item_$themeId';

  /// The third-party notices screen.
  static const String licensesScreen = 'sample_licenses_screen';

  /// One notice row, suffixed with its package name.
  static String licenseRow(String package) =>
      'sample_license_row_${_idSafe(package)}';

  /// The expanded licence body, suffixed with its package name.
  static String licenseText(String package) =>
      'sample_license_text_${_idSafe(package)}';

  /// What the screen says when the notices did not ship.
  static const String licensesEmpty = 'sample_licenses_empty';

  /// The selection bar shown instead of the nav bar.
  static const String selectionBar = 'sample_selection_bar';

  /// The "n selected" text, its own node so a flow asserts equality rather than parsing prose.
  static const String selectionCount = 'sample_selection_count';

  /// The remove action in the selection bar; the copy reads Hide from List.
  static const String selectionRemove = 'sample_selection_remove';

  /// Every id declared here, which the spec test checks against `spec/test-ids.json`.
  static const List<String> all = <String>[
    navProducts,
    navVerifications,
    navSettings,
    navToken,
    sessionCard,
    sessionCountdown,
    sessionEndedBanner,
    tokenFloat,
    tokenManualEntry,
    tokenPaste,
    tokenSimulate,
    scanTokenScreen,
    productsScreen,
    profileAvatarButton,
    settingsScreen,
    profileSummary,
    settingEnhancedSmartSelfie,
    settingAgentMode,
    settingDarkMode,
    settingConsentStep,
    settingInstructionsStep,
    settingPreviewStep,
    settingNav,
    signOut,
    signOutConfirm,
    versionLabel,
    toast,
    toastUndo,
    jobRowStatus,
    selectToggle,
    verificationDetailsScreen,
    verificationsScreen,
    verificationsEmpty,
    statusBadge,
    detailsDelete,
    detailsRefresh,
    detailsEmpty,
    profilesScreen,
    createProfile,
    profileConfigScreen,
    profileConfigCallbackUrl,
    profileConfigSave,
    profileConfigName,
    profileConfigDelete,
    profileDeleteConfirm,
    resultCard,
    resultActiveScenario,
    resultActiveTheme,
    resultRoute,
    resultEnvironment,
    resultJobId,
    resultUserId,
    resultJobStatus,
    resultResultCount,
    resultRefreshCount,
    resultLastError,
    resultSdkVersion,
    profileSwitchSheet,
    profileSwitchNew,
    newProfileSheet,
    newProfileName,
    newProfileFirstName,
    newProfileLastName,
    newProfileEmail,
    newProfilePhone,
    newProfileSave,
    userDetailsScreen,
    userDetailsHint,
    rememberDetailsSwitch,
    userDetailsProfile,
    userDetailsContinue,
    kycFormScreen,
    countryTrigger,
    idTypeTrigger,
    idNumberInput,
    kycContinue,
    countrySheet,
    countrySearch,
    countryEmpty,
    idTypeSheet,
    idTypeSearch,
    idTypeEmpty,
    scenarioDrawer,
    scenarioDrawerButton,
    licensesScreen,
    licensesEmpty,
    selectionBar,
    selectionCount,
    selectionRemove,
  ];
}
